clarity/src/vm/clarity_wasm.rs

```diff
@@ -0,0 +1,1038 @@
+                            .get_public_function_type(&function_name)
+                            .ok_or(Error::Unchecked(CheckErrors::UnknownFunction(
+                                function_name.clone(),
+                            )))?,
+                    ),
+                    2 => (
+                        DefineType::Private,
+                        caller
+                            .data()
+                            .contract_analysis
+                            .get_private_function(&function_name)
+                            .ok_or(Error::Unchecked(CheckErrors::UnknownFunction(
+                                function_name.clone(),
+                            )))?,
+                    ),
+                    _ => Err(Error::Wasm(WasmError::InvalidFunctionKind(kind)))?,
+                };
+
+                let fixed_type = match function_type {
+                    FunctionType::Fixed(fixed_type) => fixed_type,
+                    _ => Err(Error::Unchecked(CheckErrors::DefineFunctionBadSignature))?,
+                };
+
+                let function = DefinedFunction::new(
+                    fixed_type
+                        .args
+                        .iter()
+                        .map(|arg| (arg.name.clone(), arg.signature.clone()))
+                        .collect(),
+                    // TODO: We don't actually need the body here, so we
+                    // should be able to remove it. For now, this is a
+                    // placeholder.
+                    SymbolicExpression::literal_value(Value::Int(0)),
+                    define_type,
+                    &function_cname,
+                    &caller
+                        .data()
+                        .run_context
+                        .contract_context
+                        .contract_identifier
+                        .to_string(),
+                    fixed_type.returns.clone(),
+                );
+
+                // Insert this function into the context
+                caller
+                    .data_mut()
+                    .run_context
+                    .contract_context
+                    .functions
+                    .insert(function_cname, function);
+
+                Ok(())
+            },
+        )
+        .map(|_| ())
+        .map_err(|e| {
+            Error::Wasm(WasmError::UnableToLinkHostFunction(
+                "define_function".to_string(),
+                e,
+            ))
+        })
+}
+
+/// When in run-mode (not initialize-mode), this should never be called.
+fn link_define_function_fn_error<T>(linker: &mut Linker<T>) -> Result<(), Error>
+where
+    T: ClarityWasmContext,
+{
+    linker
+        .func_wrap(
+            "clarity",
+            "define_function",
+            |mut _caller: Caller<'_, T>, _kind: i32, _name_offset: i32, _name_length: i32| {
+                // FIXME: I don't understand why I have to write this like this
```

Comment:
Yes I do, it's a bit tricky.
The closure here should implement `IntoFunc`. For the arguments, no issue, but for the return type, here you are returning a `Result`, which does not implement `WasmRet`.
What does implement `WasmRet` however is `WasmTime::Result`, which they say is a wrapper for `Result<T, anyhow::Error`. 
So if I add `anyhow` as dep and rewrite the closure like this, it works:
```rust
|mut _caller: Caller<'_, T>, _kind: i32, _name_offset: i32, _name_length: i32| {
    Err::<(), _>(anyhow!(Error::Wasm(WasmError::DefineFunctionCalledInRunMode)))
}
```

Why it worked with a `?` is because this calls an implicit "into" on the error, converting it automatically to a `anyhow::Error`.